### ISSUE TITLE
Clean sessions last longer than connection

### ISSUE THREAD

**Issue by sunib** [rabbitmq/rabbitmq-mqtt#37]
I just spent some time debugging an issue with the newest version of RabbitMQ (3.5.6), turns out that my queues are not automatically deleted anymore. I'm opening MQTT connections with clean session = 1 and no client id. I'm subscribing with QoS 1.

Several reconnects of my application results in a lot of (bound!) queues that are never deleted anymore. And because I'm not using a client id I can never connect to them again (at least not with MQTT).

So what I'm going to do is just use QoS 0, way better in my use case.

But what I learned (and wanted to share):
* This was only recently introduced: https://github.com/rabbitmq/rabbitmq-mqtt/issues/30
* The standard say something about this in this chapter: http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Ref362965194

It states:
> If CleanSession is set to 1, the Client and Server MUST discard any previous Session and start a new one. **This Session lasts as long as the Network Connection.** State data associated with this Session MUST NOT be reused in any subsequent Session

So I don't think that the queues should be durable in this case. In my opinion it would be a better fit to have it like it was before issue 30 was implemented.

Any other thoughts on this one? I don't see why the MQTT protocol allows the combination anyway.

**Comment by michaelklishin** [comment]
The spec is ambiguous. How can a server know about earlier sessions?

Unfortunately some readers interpret the spec differently and do not want QoS1 queued deleted between reconnections.

We need to investigate some more but your use case is what QoS 0 is for: why store anything on disk if you want to throw it away upon disconnect?

**Comment by michaelklishin** [comment]
OK, so I think what we should do is revert #30 but also make sure that if `clean session` is disabled, we use a durable queue.

This would satisfy

```
If CleanSession is set to 0, the Server MUST resume communications with the Client based on state from the current Session (as identified by the Client identifier). If there is no Session associated with the Client identifier the Server MUST create a new Session. The Client and Server MUST store the Session after the Client and Server are disconnected [MQTT-3.1.2-4]. After the disconnection of a Session that had CleanSession set to 0, the Server MUST store further QoS 1 and QoS 2 messages that match any subscriptions that the client had at the time of disconnection as part of the Session state [MQTT-3.1.2-5].
```

Also note that the above does not address the case when client ID is not provided. How can the server know what it should be, and locate previous session state? I'm not sure.

**Comment by sunib** [comment]
Thanks for your fast answers! 

Regarding your last line, according to the spec you can just drop the connection on that combination:

> If the Client supplies a zero-byte ClientId with CleanSession set to 0, the Server MUST respond to the CONNECT Packet with a CONNACK return code 0x02 (Identifier rejected) and then close the Network Connection [MQTT-3.1.3-8].

My current knowledge brings me to the following expected behavior:

client id | clean session | action
--- | --- | ---
null | false | close connection
filled | false | create or reload durable queue
null | true | queue with auto delete
filled | true | queue with auto delete

**Comment by michaelklishin** [comment]
Ah, that makes more sense. OK, will be in 3.5.7.

**Comment by michaelklishin** [comment]
I've leant that some clients set clean session to true by default, e.g. Paho Python. That's pretty unfortunate for those trying QoS 1 with them at first.

**Comment by michaelklishin** [comment]
Fixed in #38. The case where client id is blank and clean session is set to `false` is already handled.
